File: sdk/typescript/src/ports/conformance.ts
/**
 * Conformance harness for {@link SystemPort} implementations.
 *
 * `tests/contract/systemPortContract.ts` already exercises adapters from
 * inside this package's own test suite, but it isn't shipped in `dist` — an
 * out-of-tree adapter package can't import it. This module is production
 * code so any `SystemPort` implementor can assert the same contract from
 * its own test suite.
 *
 * The port speaks CIF only (see the `SystemPort` module docs' "Canonical-
 * only" section): `upsert` receives a document containing just the fields
 * the two systems share. Fields the CIF mapping never sees are per-system
 * local state, and `upsert` must read-modify-write the shared paths onto
 * the existing native record rather than replace it wholesale — otherwise
 * a push silently deletes native-only data. Checking that needs a peek at
 * the adapter's native storage the port itself never exposes, hence
 * {@link RawAccess}: a tiny side door the adapter's own test supplies.
 *
 * Port of `core/src/ports/conformance.rs`.
 */

import { SyncError } from "../domain/error.js";
import { idempotencyKey } from "../domain/idempotency.js";
import type { JsonValue } from "../domain/types.js";
import type { ExternalRef, SystemPort } from "./system.js";

/**
 * Adapter-supplied side door into native storage, bypassing the CIF-only
 * `SystemPort` surface. Only needed for the preservation check (C1) — the
 * other checks drive the port directly.
 */
export interface RawAccess {
  /**
   * Seed a raw native record directly in the adapter's backing store,
   * bypassing `upsert`. `native` may contain fields the CIF mapping never
   * touches — the harness uses that to prove a push doesn't clobber them.
   */
  seedRaw(entityType: string, canonicalId: string, native: JsonValue): Promise<ExternalRef>;

  /**
   * Read the native record back, bypassing whatever CIF projection the
   * port applies on `fetch`.
   */
  readRaw(entityType: string, canonicalId: string): Promise<JsonValue>;
}

/**
 * Assert a {@link SystemPort} implementation honors the port's documented
 * contract. Throws (with a message naming the violated rule) on the first
 * failing check.
 */
export async function assertSystemPortContract(port: SystemPort, raw: RawAccess): Promise<void> {
  await assertC1Preservation(port, raw);
  await assertC2VersionGuard(port);
  await assertC3Idempotency(port);
}

/**
 * C1 — preservation. `upsert` must update the shared CIF paths on the
 * existing native record and leave everything else alone (read-modify-
 * write, not replace) — see this module's doc comment.
 */
async function assertC1Preservation(port: SystemPort, raw: RawAccess): Promise<void> {
  const entityType = "conformance_entity";
  const canonicalId = "conformance-c1";
  const seeded = await raw.seedRaw(entityType, canonicalId, {
    shared: "old",
    native_only: "keep-me",
  });

  const cif: JsonValue = { shared: "new" };
  const ik = idempotencyKey(canonicalId, "upsert", cif);
  await port.upsert(entityType, canonicalId, cif, seeded.version, ik);

  const after = await raw.readRaw(entityType, canonicalId);
  const afterObj = after as Record<string, JsonValue>;
  if (afterObj.shared !== "new") {
    throw new Error("C1 preservation violated: upsert must update the shared path it was given");
  }
  if (afterObj.native_only !== "keep-me") {
    throw new Error(
      "C1 preservation violated: upsert must read-modify-write the native record, " +
        "not replace it — a native-only field vanished after a push",
    );
  }
}

/**
 * C2 — version guard. The port's "Optimistic concurrency" docs promise a
 * mismatched `expectVersion` throws `SyncError` with `kind: "StaleWrite"`
 * rather than silently overwriting, so every `SystemPort` must honor it —
 * this check always runs (no skip condition).
 */
async function assertC2VersionGuard(port: SystemPort): Promise<void> {
  const entityType = "conformance_entity";
  const canonicalId = "conformance-c2";
  const cif: JsonValue = { shared: "v1" };
  const ik1 = idempotencyKey(canonicalId, "upsert", cif);
  const created = await port.upsert(entityType, canonicalId, cif, undefined, ik1);

  const stale = `stale-${created.version ?? "none"}`;
  const cif2: JsonValue = { shared: "v2" };
  const ik2 = idempotencyKey(canonicalId, "upsert", cif2);

  let threw: unknown;
  try {
    await port.upsert(entityType, canonicalId, cif2, stale, ik2);
  } catch (e) {
    threw = e;
  }
  if (threw === undefined) {
    throw new Error(
      "C2 version guard violated: upsert with a mismatched expectVersion must be rejected, not silently applied",
    );
  }
  if (!(threw instanceof SyncError) || threw.kind !== "StaleWrite") {
    throw new Error(
      `C2 version guard violated: expected SyncError with kind "StaleWrite" on version mismatch, got ${String(threw)}`,
    );
  }
}

/**
 * C3 — idempotency. The port's "Idempotency" docs require adapters to
 * dedup a repeated `idempotencyKey` (via the external system's mechanism
 * or a table of their own), so every `SystemPort` must honor it — this
 * check always runs (no skip condition).
 */
async function assertC3Idempotency(port: SystemPort): Promise<void> {
  const entityType = "conformance_entity";
  const canonicalId = "conformance-c3";
  const cif: JsonValue = { shared: "same" };
  const ik = idempotencyKey(canonicalId, "upsert", cif);

  const r1 = await port.upsert(entityType, canonicalId, cif, undefined, ik);
  const r2 = await port.upsert(entityType, canonicalId, cif, undefined, ik);

  if (r1.system !== r2.system || r1.externalId !== r2.externalId || r1.version !== r2.version) {
    throw new Error(
      "C3 idempotency violated: a repeat upsert with the same idempotency key must be a no-op, not create a new write",
    );
  }
}

File: core/src/ports/conformance.rs
//! Conformance harness for [`SystemPort`] implementations.
//!
//! `core/tests/integration/contract_tests.rs` already exercises the
//! reference adapter, but a `tests/` file isn't part of the published
//! crate — an out-of-tree adapter crate can't call into it. This module is
//! library code so any `SystemPort` implementor can assert the same
//! contract from its own test suite.
//!
//! The trait speaks CIF only (see the [`crate::ports::system`] module
//! docs' "Canonical-only" section): `upsert` receives a document
//! containing just the fields the two systems share. Fields the CIF
//! mapping never sees are per-system local state, and `upsert` must
//! read-modify-write the shared paths onto the existing native record
//! rather than replace it wholesale — otherwise a push silently deletes
//! native-only data. Checking that needs a peek at the adapter's native
//! storage that the port itself never exposes, hence [`RawAccess`]: a tiny
//! side door the adapter's own test supplies.

use crate::domain::error::SyncError;
use crate::domain::idempotency::idempotency_key;
use crate::ports::system::{ExternalRef, SystemPort};
use async_trait::async_trait;
use serde_json::{Value, json};

/// Adapter-supplied side door into native storage, bypassing the CIF-only
/// `SystemPort` surface. Only needed for the preservation check (C1) —
/// the other checks drive the port directly.
#[async_trait]
pub trait RawAccess {
    /// Seed a raw native record directly in the adapter's backing store,
    /// bypassing `upsert`. `native` may contain fields the CIF mapping
    /// never touches — the harness uses that to prove a push doesn't
    /// clobber them.
    async fn seed_raw(&self, entity_type: &str, canonical_id: &str, native: Value) -> ExternalRef;

    /// Read the native record back, bypassing whatever CIF projection the
    /// port applies on `fetch`.
    async fn read_raw(&self, entity_type: &str, canonical_id: &str) -> Value;
}

/// Assert a [`SystemPort`] implementation honors the trait's documented
/// contract. Panics (with a message naming the violated rule) on the first
/// failing check.
pub async fn assert_system_port_contract<P: SystemPort, R: RawAccess>(port: &P, raw: &R) {
    assert_c1_preservation(port, raw).await;
    assert_c2_version_guard(port).await;
    assert_c3_idempotency(port).await;
}

/// C1 — preservation. `upsert` must update the shared CIF paths on the
/// existing native record and leave everything else alone (read-modify-
/// write, not replace) — see this module's doc comment.
async fn assert_c1_preservation<P: SystemPort, R: RawAccess>(port: &P, raw: &R) {
    let entity_type = "conformance_entity";
    let canonical_id = "conformance-c1";
    let seeded = raw
        .seed_raw(
            entity_type,
            canonical_id,
            json!({"shared": "old", "native_only": "keep-me"}),
        )
        .await;

    let cif = json!({"shared": "new"});
    let ik = idempotency_key(canonical_id, "upsert", &cif);
    port.upsert(entity_type, canonical_id, &cif, seeded.version.as_deref(), &ik)
        .await
        .expect("C1 preservation: upsert of a shared-only CIF document must succeed");

    let after = raw.read_raw(entity_type, canonical_id).await;
    assert_eq!(
        after.get("shared"),
        Some(&json!("new")),
        "C1 preservation violated: upsert must update the shared path it was given"
    );
    assert_eq!(
        after.get("native_only"),
        Some(&json!("keep-me")),
        "C1 preservation violated: upsert must read-modify-write the native record, \
         not replace it — a native-only field vanished after a push"
    );
}

/// C2 — version guard. The trait's "Optimistic concurrency" docs promise a
/// mismatched `expect_version` returns [`SyncError::StaleWrite`] rather
/// than silently overwriting, so every `SystemPort` must honor it — this
/// check always runs (no skip condition).
async fn assert_c2_version_guard<P: SystemPort>(port: &P) {
    let entity_type = "conformance_entity";
    let canonical_id = "conformance-c2";
    let cif = json!({"shared": "v1"});
    let ik1 = idempotency_key(canonical_id, "upsert", &cif);
    let created = port
        .upsert(entity_type, canonical_id, &cif, None, &ik1)
        .await
        .expect("C2 version guard: initial upsert must succeed");

    let stale = format!("stale-{}", created.version.as_deref().unwrap_or("none"));
    let cif2 = json!({"shared": "v2"});
    let ik2 = idempotency_key(canonical_id, "upsert", &cif2);
    let err = port
        .upsert(entity_type, canonical_id, &cif2, Some(&stale), &ik2)
        .await
        .expect_err(
            "C2 version guard violated: upsert with a mismatched expect_version \
             must be rejected, not silently applied",
        );
    assert!(
        matches!(err, SyncError::StaleWrite { .. }),
        "C2 version guard violated: expected SyncError::StaleWrite on version mismatch, got {err:?}"
    );
}

/// C3 — idempotency. The trait's "Idempotency" docs require adapters to
/// dedup a repeated `idempotency_key` (via the external system's mechanism
/// or a table of their own), so every `SystemPort` must honor it — this
/// check always runs (no skip condition).
async fn assert_c3_idempotency<P: SystemPort>(port: &P) {
    let entity_type = "conformance_entity";
    let canonical_id = "conformance-c3";
    let cif = json!({"shared": "same"});
    let ik = idempotency_key(canonical_id, "upsert", &cif);

    let r1 = port
        .upsert(entity_type, canonical_id, &cif, None, &ik)
        .await
        .expect("C3 idempotency: first upsert must succeed");
    let r2 = port
        .upsert(entity_type, canonical_id, &cif, None, &ik)
        .await
        .expect("C3 idempotency: repeat upsert with same idempotency key must succeed");

    assert_eq!(
        r1, r2,
        "C3 idempotency violated: a repeat upsert with the same idempotency key \
         must be a no-op, not create a new write"
    );
}
